casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -143,11 +144,11 @@ class RuntimeManager private (val emptyStateHash: ByteString, runtimeContainer:
       terms match {
         case deploy +: rem =>
           runtime.space.reset(hash)
-          val availablePhlos             = CostAccount(Integer.MAX_VALUE)
-          implicit val costAccountingAlg = CostAccountingAlg.unsafe[Task](availablePhlos) //FIXME this needs to come from the deploy params
-          val (phlosLeft, errors)        = injAttempt(deploy, runtime.reducer, runtime.errorLog)
-          val cost                       = phlosLeft.copy(cost = availablePhlos.cost.value - phlosLeft.cost)
-          val newCheckpoint              = runtime.space.createCheckpoint()
+          val availablePhlos = Cost(Integer.MAX_VALUE)
+          Await.ready(runtime.reducer.setAvailablePhlos(availablePhlos).runAsync, 1.second) // FIXME: This needs to come from the deploy params
```

Comment:
I am very wary of the `Await` in here. I'm also not yet sure why we can't avoid mutating the reducer's state. Please provide a high level overview of what exactly you're trying to achieve and why a cleaner approach is not possible.

Reply:
Because when we leave RSpace with a matching data in `Dispatcher` we need to continue burning the same phlos. If we have a `costAccountingAlg` dependency at method call level this won't work. We already share same `Reduce` instance in `Dispatch` so it makes sense to move the state inside to `Reduce`. Since `costAccountingAlg` is not required at method call anymore, we need a way to set new `phloLimit` which is known only when we know deploy data.